README.md
# ocaml-win-error: manipulate Windows system errors

If your application is likely to run on Windows, simply replace uses of
`Unix.error_message` with `Win_error.error_message` for enhanced error
reporting. Note this code is platform-agnostic so can be linked into Unix
applications too.

A simple example:
```ocaml
try
  let fd = Unix.openfile "/tmp/foo" [ Unix.O_RDONLY ] 0 in
  Unix.close fd
with
| Unix.Unix_error(e, _, _) ->
  Printf.fprintf stderr "Caught: %s\n%!" (Win_error.error_message e)
```

## What's going on?

Errors are typically first thrown by the OCaml standard library as `Unix.Unix_error`
exceptions. In some cases these are mapped onto their Unix equivalents on
such as `Unix.ENOENT`, but when the code isn't recognised OCaml will raise
`Unix.EUNKNOWNERR (-code)`. This library is able to convert these codes back
into human-readable strings.


CHANGES.md
1.0

- Improve ocamldoc for the library (@avsm)
- Port to dune (@avsm)
- Install toplevel printers (@avsm)
- Upgrade opam metadata to 2.0 format (@avsm)

0.3 (2018-06-14)
- Port to jbuilder

0.2 (2016-05-06)
- Enable build on OCaml 3.12.1
- `Win_error.error_message` now behaves differently depending on
  `Sys.os_type`

0.1 (2016-05-05)
- Initial release


lib/win_error_t.ml

type t =
  | Success
  | Invalid_function
  | File_not_found
  | Path_not_found
  | Too_many_open_files
  | Access_denied
  | Invalid_handle
  | Arena_trashed
  | Not_enough_memory
  | Invalid_block
  | Bad_environment
  | Bad_format
  | Invalid_access
  | Invalid_data
  | Outofmemory
  | Invalid_drive
  | Current_directory
  | Not_same_device
  | No_more_files
  | Write_protect
  | Bad_unit
  | Not_ready
  | Bad_command
  | Crc
  | Bad_length
  | Seek
  | Not_dos_disk
  | Sector_not_found
  | Out_of_paper
  | Write_fault
  | Read_fault
  | Gen_failure
  | Sharing_violation
  | Lock_violation
  | Wrong_disk
  | Sharing_buffer_exceeded
  | Handle_eof
  | Handle_disk_full
  | Not_supported
  | Rem_not_list
  | Dup_name
  | Bad_netpath
  | Network_busy
  | Dev_not_exist
  | Too_many_cmds
  | Adap_hdw_err
  | Bad_net_resp
  | Unexp_net_err
  | Bad_rem_adap
  | Printq_full
  | No_spool_space
  | Print_cancelled
  | Netname_deleted
  | Network_access_denied
  | Bad_dev_type
  | Bad_net_name
  | Too_many_names
  | Too_many_sess
  | Sharing_paused
  | Req_not_accep
  | Redir_paused
  | File_exists
  | Cannot_make
  | Fail_I24
  | Out_of_structures
  | Already_assigned
  | Invalid_password
  | Invalid_parameter
  | Net_write_fault
  | No_proc_slots
  | Too_many_semaphores
  | Excl_sem_already_owned
  | Sem_is_set
  | Too_many_sem_requests
  | Invalid_at_interrupt_time
  | Sem_owner_died
  | Sem_user_limit
  | Disk_change
  | Drive_locked
  | Broken_pipe
  | Open_failed
  | Buffer_overflow
  | Disk_full
  | No_more_search_handles
  | Invalid_target_handle
  | Invalid_category
  | Invalid_verify_switch
  | Bad_driver_level
  | Call_not_implemented
  | Sem_timeout
  | Insufficient_buffer
  | Invalid_name
  | Invalid_level
  | No_volume_label
  | Mod_not_found
  | Proc_not_found
  | Wait_no_children
  | Child_not_complete
  | Direct_access_handle
  | Negative_seek
  | Seek_on_device
  | Is_join_target
  | Is_joined
  | Is_substed
  | Not_joined
  | Not_substed
  | Join_to_join
  | Subst_to_subst
  | Join_to_subst
  | Subst_to_join
  | Busy_drive
  | Same_drive
  | Dir_not_root
  | Dir_not_empty
  | Is_subst_path
  | Is_join_path
  | Path_busy
  | Is_subst_target
  | System_trace
  | Invalid_event_count
  | Too_many_muxwaiters
  | Invalid_list_format
  | Label_too_long
  | Too_many_tcps
  | Signal_refused
  | Discarded
  | Not_locked
  | Bad_threadid_addr
  | Bad_arguments
  | Bad_pathname
  | Signal_pending
  | Max_thrds_reached
  | Lock_failed
  | Busy
  | Device_support_in_progress
  | Cancel_violation
  | Atomic_locks_not_supported
  | Invalid_segment_number
  | Invalid_ordinal
  | Already_exists
  | Invalid_flag_number
  | Sem_not_found
  | Error_invalid_starting_codeseg
  | Invalid_stackseg
  | Invalid_moduletype
  | Invalid_exe_signature
  | Exe_marked_invalid
  | Bad_exe_format
  | Iterated_data_exceeds_64k
  | Invalid_minallocsize
  | Dynlink_from_invalid_ring
  | Iopl_not_enabled
  | Invalid_segdpl
  | Autodataseg_exceeds_64k
  | Ring2seg_must_be_movable
  | Reloc_chain_xeeds_seglim
  | Infloop_in_reloc_chain
  | Envvar_not_found
  | No_signal_sent
  | Filename_exced_range
  | Ring2_stack_in_use
  | Meta_expansion_too_long
  | Invalid_signal_number
  | Thread_1_inactive
  | Lock_failed
  | Too_many_modules
  | Nesting_not_allowed
  | Exe_machine_type_mismatch
  | Exe_cannot_modify_signed_binary
  | Exe_cannot_modify_strong_signed_binary
  | File_checked_out
  | Checkout_required
  | Bad_file_type
  | File_too_large
  | Forms_auth_required
  | Virus_infected
  | Virus_deleted
  | Pipe_local
  | Bad_pipe
  | Pipe_busy
  | No_data
  | Pipe_not_connected
  | More_data
  | Vc_disconnected
  | Invalid_ea_name
  | Ea_list_inconsistent
  | Wait_timeout
  | No_more_items
  | Cannot_copy
  | Directory
  | Eas_didnt_fit
  | Ea_file_corrupt
  | Ea_table_full
  | Invalid_ea_handle
  | Eas_not_supported
  | Not_owner
  | Too_many_posts
  | Partial_copy
  | Oplock_not_granted
  | Invalid_oplock_protocol
  | Disk_too_fragmented
  | Delete_pending
  | Incompatible_with_global_short_name_registry_setting
  | Short_names_not_enabled_on_volume
  | Security_stream_is_inconsistent
  | Invalid_lock_range
  | Image_subsystem_not_present
  | Notification_guid_already_defined
  | Invalid_exception_handler
  | Duplicate_privileges
  | No_ranges_processed
  | Not_allowed_on_system_file
  | Disk_resources_exhausted
  | Invalid_token
  | Device_feature_not_supported
  | Mr_mid_not_found
  | Scope_not_found
  | Undefined_scope
  | Invalid_cap
  | Device_unreachable
  | Device_no_resources
  | Data_checksum_error
  | Intermixed_kernel_ea_operation
  | File_level_trim_not_supported
  | Offset_alignment_violation
  | Invalid_field_in_parameter_list
  | Operation_in_progress
  | Bad_device_path
  | Too_many_descriptors
  | Scrub_data_disabled
  | Not_redundant_storage
  | Resident_file_not_supported
  | Compressed_file_not_supported
  | Directory_not_supported
  | Not_read_from_copy
  | Fail_noaction_reboot
  | Fail_shutdown
  | Fail_restart
  | Max_sessions_reached
  | Thread_mode_already_background
  | Thread_mode_not_background
  | Process_mode_already_background
  | Process_mode_not_background
  | Invalid_address


lib/win_error_of_int.ml

let of_int = function
  | 0 -> Some (Success)
  | 1 -> Some (Invalid_function)
  | 2 -> Some (File_not_found)
  | 3 -> Some (Path_not_found)
  | 4 -> Some (Too_many_open_files)
  | 5 -> Some (Access_denied)
  | 6 -> Some (Invalid_handle)
  | 7 -> Some (Arena_trashed)
  | 8 -> Some (Not_enough_memory)
  | 9 -> Some (Invalid_block)
  | 10 -> Some (Bad_environment)
  | 11 -> Some (Bad_format)
  | 12 -> Some (Invalid_access)
  | 13 -> Some (Invalid_data)
  | 14 -> Some (Outofmemory)
  | 15 -> Some (Invalid_drive)
  | 16 -> Some (Current_directory)
  | 17 -> Some (Not_same_device)
  | 18 -> Some (No_more_files)
  | 29 -> Some (Write_protect)
  | 20 -> Some (Bad_unit)
  | 21 -> Some (Not_ready)
  | 22 -> Some (Bad_command)
  | 23 -> Some (Crc)
  | 24 -> Some (Bad_length)
  | 25 -> Some (Seek)
  | 26 -> Some (Not_dos_disk)
  | 27  -> Some (Sector_not_found)
  | 28 -> Some (Out_of_paper)
  | 29 -> Some (Write_fault)
  | 30 -> Some (Read_fault)
  | 31 -> Some (Gen_failure)
  | 32 -> Some (Sharing_violation)
  | 33 -> Some (Lock_violation)
  | 34 -> Some (Wrong_disk)
  | 36 -> Some (Sharing_buffer_exceeded)
  | 38 -> Some (Handle_eof)
  | 39 -> Some (Handle_disk_full)
  | 50 -> Some (Not_supported)
  | 51 -> Some (Rem_not_list)
  | 52 -> Some (Dup_name)
  | 53 -> Some (Bad_netpath)
  | 54 -> Some (Network_busy)
  | 55 -> Some (Dev_not_exist)
  | 56 -> Some (Too_many_cmds)
  | 57 -> Some (Adap_hdw_err)
  | 58 -> Some (Bad_net_resp)
  | 59 -> Some (Unexp_net_err)
  | 60 -> Some (Bad_rem_adap)
  | 61 -> Some (Printq_full)
  | 62 -> Some (No_spool_space)
  | 63 -> Some (Print_cancelled)
  | 64 -> Some (Netname_deleted)
  | 65 -> Some (Network_access_denied)
  | 66 -> Some (Bad_dev_type)
  | 67 -> Some (Bad_net_name)
  | 68 -> Some (Too_many_names)
  | 69 -> Some (Too_many_sess)
  | 70 -> Some (Sharing_paused)
  | 71 -> Some (Req_not_accep)
  | 72-> Some (Redir_paused)
  | 80 -> Some (File_exists)
  | 82 -> Some (Cannot_make)
  | 83 -> Some (Fail_I24)
  | 84 -> Some (Out_of_structures)
  | 85 -> Some (Already_assigned)
  | 86 -> Some (Invalid_password)
  | 87 -> Some (Invalid_parameter)
  | 88 -> Some (Net_write_fault)
  | 89 -> Some (No_proc_slots)
  | 100 -> Some (Too_many_semaphores)
  | 101 -> Some (Excl_sem_already_owned)
  | 102 -> Some (Sem_is_set)
  | 103 -> Some (Too_many_sem_requests)
  | 104 -> Some (Invalid_at_interrupt_time)
  | 105 -> Some (Sem_owner_died)
  | 106 -> Some (Sem_user_limit)
  | 107 -> Some (Disk_change)
  | 108 -> Some (Drive_locked)
  | 109 -> Some (Broken_pipe)
  | 110 -> Some (Open_failed)
  | 111 -> Some (Buffer_overflow)
  | 112 -> Some (Disk_full)
  | 113 -> Some (No_more_search_handles)
  | 114 -> Some (Invalid_target_handle)
  | 117 -> Some (Invalid_category)
  | 118 -> Some (Invalid_verify_switch)
  | 119 -> Some (Bad_driver_level)
  | 120 -> Some (Call_not_implemented)
  | 121 -> Some (Sem_timeout)
  | 122 -> Some (Insufficient_buffer)
  | 123 -> Some (Invalid_name)
  | 124 -> Some (Invalid_level)
  | 125 -> Some (No_volume_label)
  | 126 -> Some (Mod_not_found)
  | 127 -> Some (Proc_not_found)
  | 128 -> Some (Wait_no_children)
  | 129 -> Some (Child_not_complete)
  | 130 -> Some (Direct_access_handle)
  | 131 -> Some (Negative_seek)
  | 132 -> Some (Seek_on_device)
  | 133 -> Some (Is_join_target)
  | 134 -> Some (Is_joined)
  | 135 -> Some (Is_substed)
  | 136 -> Some (Not_joined)
  | 137 -> Some (Not_substed)
  | 138 -> Some (Join_to_join)
  | 139 -> Some (Subst_to_subst)
  | 140 -> Some (Join_to_subst)
  | 141 -> Some (Subst_to_join)
  | 142 -> Some (Busy_drive)
  | 143 -> Some (Same_drive)
  | 144 -> Some (Dir_not_root)
  | 145 -> Some (Dir_not_empty)
  | 146 -> Some (Is_subst_path)
  | 147 -> Some (Is_join_path)
  | 148 -> Some (Path_busy)
  | 149 -> Some (Is_subst_target)
  | 150 -> Some (System_trace)
  | 151 -> Some (Invalid_event_count)
  | 152 -> Some (Too_many_muxwaiters)
  | 153 -> Some (Invalid_list_format)
  | 154 -> Some (Label_too_long)
  | 155 -> Some (Too_many_tcps)
  | 156 -> Some (Signal_refused)
  | 157 -> Some (Discarded)
  | 158 -> Some (Not_locked)
  | 159 -> Some (Bad_threadid_addr)
  | 160 -> Some (Bad_arguments)
  | 161 -> Some (Bad_pathname)
  | 162 -> Some (Signal_pending)
  | 164 -> Some (Max_thrds_reached)
  | 167 -> Some (Lock_failed)
  | 170 -> Some (Busy)
  | 171 -> Some (Device_support_in_progress)
  | 173 -> Some (Cancel_violation)
  | 174 -> Some (Atomic_locks_not_supported)
  | 180 -> Some (Invalid_segment_number)
  | 182 -> Some (Invalid_ordinal)
  | 183 -> Some (Already_exists)
  | 186 -> Some (Invalid_flag_number)
  | 187 -> Some (Sem_not_found)
  | 188 -> Some (Error_invalid_starting_codeseg)
  | 189 -> Some (Invalid_stackseg)
  | 190 -> Some (Invalid_moduletype)
  | 191 -> Some (Invalid_exe_signature)
  | 192 -> Some (Exe_marked_invalid)
  | 193 -> Some (Bad_exe_format)
  | 194 -> Some (Iterated_data_exceeds_64k)
  | 195 -> Some (Invalid_minallocsize)
  | 196 -> Some (Dynlink_from_invalid_ring)
  | 197 -> Some (Iopl_not_enabled)
  | 198 -> Some (Invalid_segdpl)
  | 199 -> Some (Autodataseg_exceeds_64k)
  | 200 -> Some (Ring2seg_must_be_movable)
  | 201 -> Some (Reloc_chain_xeeds_seglim)
  | 202 -> Some (Infloop_in_reloc_chain)
  | 203 -> Some (Envvar_not_found)
  | 205 -> Some (No_signal_sent)
  | 206 -> Some (Filename_exced_range)
  | 207 -> Some (Ring2_stack_in_use)
  | 208 -> Some (Meta_expansion_too_long)
  | 209 -> Some (Invalid_signal_number)
  | 210 -> Some (Thread_1_inactive)
  | 212 -> Some (Lock_failed)
  | 214 -> Some (Too_many_modules)
  | 215 -> Some (Nesting_not_allowed)
  | 216 -> Some (Exe_machine_type_mismatch)
  | 217 -> Some (Exe_cannot_modify_signed_binary)
  | 218 -> Some (Exe_cannot_modify_strong_signed_binary)
  | 220 -> Some (File_checked_out)
  | 221 -> Some (Checkout_required)
  | 222 -> Some (Bad_file_type)
  | 223 -> Some (File_too_large)
  | 224 -> Some (Forms_auth_required)
  | 225 -> Some (Virus_infected)
  | 226 -> Some (Virus_deleted)
  | 229 -> Some (Pipe_local)
  | 230 -> Some (Bad_pipe)
  | 231 -> Some (Pipe_busy)
  | 231 -> Some (No_data)
  | 233 -> Some (Pipe_not_connected)
  | 234 -> Some (More_data)
  | 240 -> Some (Vc_disconnected)
  | 254 -> Some (Invalid_ea_name)
  | 255 -> Some (Ea_list_inconsistent)
  | 258 -> Some (Wait_timeout)
  | 259 -> Some (No_more_items)
  | 266 -> Some (Cannot_copy)
  | 267 -> Some (Directory)
  | 275 -> Some (Eas_didnt_fit)
  | 276 -> Some (Ea_file_corrupt)
  | 277 -> Some (Ea_table_full)
  | 278 -> Some (Invalid_ea_handle)
  | 282 -> Some (Eas_not_supported)
  | 288 -> Some (Not_owner)
  | 298 -> Some (Too_many_posts)
  | 299 -> Some (Partial_copy)
  | 300 -> Some (Oplock_not_granted)
  | 301 -> Some (Invalid_oplock_protocol)
  | 302 -> Some (Disk_too_fragmented)
  | 303 -> Some (Delete_pending)
  | 304 -> Some (Incompatible_with_global_short_name_registry_setting)
  | 305 -> Some (Short_names_not_enabled_on_volume)
  | 306 -> Some (Security_stream_is_inconsistent)
  | 307 -> Some (Invalid_lock_range)
  | 308 -> Some (Image_subsystem_not_present)
  | 309 -> Some (Notification_guid_already_defined)
  | 310 -> Some (Invalid_exception_handler)
  | 311 -> Some (Duplicate_privileges)
  | 312 -> Some (No_ranges_processed)
  | 313 -> Some (Not_allowed_on_system_file)
  | 314 -> Some (Disk_resources_exhausted)
  | 315 -> Some (Invalid_token)
  | 316 -> Some (Device_feature_not_supported)
  | 317 -> Some (Mr_mid_not_found)
  | 318 -> Some (Scope_not_found)
  | 319 -> Some (Undefined_scope)
  | 320 -> Some (Invalid_cap)
  | 321 -> Some (Device_unreachable)
  | 322 -> Some (Device_no_resources)
  | 323 -> Some (Data_checksum_error)
  | 324 -> Some (Intermixed_kernel_ea_operation)
  | 326 -> Some (File_level_trim_not_supported)
  | 327 -> Some (Offset_alignment_violation)
  | 328 -> Some (Invalid_field_in_parameter_list)
  | 329 -> Some (Operation_in_progress)
  | 330 -> Some (Bad_device_path)
  | 331 -> Some (Too_many_descriptors)
  | 332 -> Some (Scrub_data_disabled)
  | 333 -> Some (Not_redundant_storage)
  | 334 -> Some (Resident_file_not_supported)
  | 335 -> Some (Compressed_file_not_supported)
  | 336 -> Some (Directory_not_supported)
  | 337 -> Some (Not_read_from_copy)
  | 350 -> Some (Fail_noaction_reboot)
  | 351 -> Some (Fail_shutdown)
  | 352 -> Some (Fail_restart)
  | 353 -> Some (Max_sessions_reached)
  | 400 -> Some (Thread_mode_already_background)
  | 401 -> Some (Thread_mode_not_background)
  | 402 -> Some (Process_mode_already_background)
  | 403 -> Some (Process_mode_not_background)
  | 487 -> Some (Invalid_address)
  | _ -> None


myocamlbuild.ml
(* OASIS_START *)
(* DO NOT EDIT (digest: ecba0625a15c7804668d8cae1d9b7353) *)
module OASISGettext = struct
(* # 22 "src/oasis/OASISGettext.ml" *)


  let ns_ str =
    str


  let s_ str =
    str


  let f_ (str: ('a, 'b, 'c, 'd) format4) =
    str


  let fn_ fmt1 fmt2 n =
    if n = 1 then
      fmt1^^""
    else
      fmt2^^""


  let init =
    []


end

module OASISExpr = struct
(* # 22 "src/oasis/OASISExpr.ml" *)





  open OASISGettext


  type test = string


  type flag = string


  type t =
    | EBool of bool
    | ENot of t
    | EAnd of t * t
    | EOr of t * t
    | EFlag of flag
    | ETest of test * string



  type 'a choices = (t * 'a) list


  let eval var_get t =
    let rec eval' =
      function
        | EBool b ->
            b

        | ENot e ->
            not (eval' e)

        | EAnd (e1, e2) ->
            (eval' e1) && (eval' e2)

        | EOr (e1, e2) ->
            (eval' e1) || (eval' e2)

        | EFlag nm ->
            let v =
              var_get nm
            in
              assert(v = "true" || v = "false");
              (v = "true")

        | ETest (nm, vl) ->
            let v =
              var_get nm
            in
              (v = vl)
    in
      eval' t


  let choose ?printer ?name var_get lst =
    let rec choose_aux =
      function
        | (cond, vl) :: tl ->
            if eval var_get cond then
              vl
            else
              choose_aux tl
        | [] ->
            let str_lst =
              if lst = [] then
                s_ "<empty>"
              else
                String.concat
                  (s_ ", ")
                  (List.map
                     (fun (cond, vl) ->
                        match printer with
                          | Some p -> p vl
                          | None -> s_ "<no printer>")
                     lst)
            in
              match name with
                | Some nm ->
                    failwith
                      (Printf.sprintf
                         (f_ "No result for the choice list '%s': %s")
                         nm str_lst)
                | None ->
                    failwith
                      (Printf.sprintf
                         (f_ "No result for a choice list: %s")
                         str_lst)
    in
      choose_aux (List.rev lst)


end


# 132 "myocamlbuild.ml"
module BaseEnvLight = struct
(* # 22 "src/base/BaseEnvLight.ml" *)


  module MapString = Map.Make(String)


  type t = string MapString.t


  let default_filename =
    Filename.concat
      (Sys.getcwd ())
      "setup.data"


  let load ?(allow_empty=false) ?(filename=default_filename) () =
    if Sys.file_exists filename then
      begin
        let chn =
          open_in_bin filename
        in
        let st =
          Stream.of_channel chn
        in
        let line =
          ref 1
        in
        let st_line =
          Stream.from
            (fun _ ->
               try
                 match Stream.next st with
                   | '\n' -> incr line; Some '\n'
                   | c -> Some c
               with Stream.Failure -> None)
        in
        let lexer =
          Genlex.make_lexer ["="] st_line
        in
        let rec read_file mp =
          match Stream.npeek 3 lexer with
            | [Genlex.Ident nm; Genlex.Kwd "="; Genlex.String value] ->
                Stream.junk lexer;
                Stream.junk lexer;
                Stream.junk lexer;
                read_file (MapString.add nm value mp)
            | [] ->
                mp
            | _ ->
                failwith
                  (Printf.sprintf
                     "Malformed data file '%s' line %d"
                     filename !line)
        in
        let mp =
          read_file MapString.empty
        in
          close_in chn;
          mp
      end
    else if allow_empty then
      begin
        MapString.empty
      end
    else
      begin
        failwith
          (Printf.sprintf
             "Unable to load environment, the file '%s' doesn't exist."
             filename)
      end


  let rec var_expand str env =
    let buff =
      Buffer.create ((String.length str) * 2)
    in
      Buffer.add_substitute
        buff
        (fun var ->
           try
             var_expand (MapString.find var env) env
           with Not_found ->
             failwith
               (Printf.sprintf
                  "No variable %s defined when trying to expand %S."
                  var
                  str))
        str;
      Buffer.contents buff


  let var_get name env =
    var_expand (MapString.find name env) env


  let var_choose lst env =
    OASISExpr.choose
      (fun nm -> var_get nm env)
      lst
end


# 237 "myocamlbuild.ml"
module MyOCamlbuildFindlib = struct
(* # 22 "src/plugins/ocamlbuild/MyOCamlbuildFindlib.ml" *)


  (** OCamlbuild extension, copied from
    * http://brion.inria.fr/gallium/index.php/Using_ocamlfind_with_ocamlbuild
    * by N. Pouillard and others
    *
    * Updated on 2009/02/28
    *
    * Modified by Sylvain Le Gall
    *)
  open Ocamlbuild_plugin

  type conf =
    { no_automatic_syntax: bool;
    }

  (* these functions are not really officially exported *)
  let run_and_read =
    Ocamlbuild_pack.My_unix.run_and_read


  let blank_sep_strings =
    Ocamlbuild_pack.Lexers.blank_sep_strings


  let exec_from_conf exec =
    let exec =
      let env_filename = Pathname.basename BaseEnvLight.default_filename in
      let env = BaseEnvLight.load ~filename:env_filename ~allow_empty:true () in
      try
        BaseEnvLight.var_get exec env
      with Not_found ->
        Printf.eprintf "W: Cannot get variable %s\n" exec;
        exec
    in
    let fix_win32 str =
      if Sys.os_type = "Win32" then begin
        let buff = Buffer.create (String.length str) in
        (* Adapt for windowsi, ocamlbuild + win32 has a hard time to handle '\\'.
         *)
        String.iter
          (fun c -> Buffer.add_char buff (if c = '\\' then '/' else c))
          str;
        Buffer.contents buff
      end else begin
        str
      end
    in
      fix_win32 exec

  let split s ch =
    let buf = Buffer.create 13 in
    let x = ref [] in
    let flush () =
      x := (Buffer.contents buf) :: !x;
      Buffer.clear buf
    in
      String.iter
        (fun c ->
           if c = ch then
             flush ()
           else
             Buffer.add_char buf c)
        s;
      flush ();
      List.rev !x


  let split_nl s = split s '\n'


  let before_space s =
    try
      String.before s (String.index s ' ')
    with Not_found -> s

  (* ocamlfind command *)
  let ocamlfind x = S[Sh (exec_from_conf "ocamlfind"); x]

  (* This lists all supported packages. *)
  let find_packages () =
    List.map before_space (split_nl & run_and_read (exec_from_conf "ocamlfind" ^ " list"))


  (* Mock to list available syntaxes. *)
  let find_syntaxes () = ["camlp4o"; "camlp4r"]


  let well_known_syntax = [
    "camlp4.quotations.o";
    "camlp4.quotations.r";
    "camlp4.exceptiontracer";
    "camlp4.extend";
    "camlp4.foldgenerator";
    "camlp4.listcomprehension";
    "camlp4.locationstripper";
    "camlp4.macro";
    "camlp4.mapgenerator";
    "camlp4.metagenerator";
    "camlp4.profiler";
    "camlp4.tracer"
  ]


  let dispatch conf =
    function
      | After_options ->
          (* By using Before_options one let command line options have an higher
           * priority on the contrary using After_options will guarantee to have
           * the higher priority override default commands by ocamlfind ones *)
          Options.ocamlc     := ocamlfind & A"ocamlc";
          Options.ocamlopt   := ocamlfind & A"ocamlopt";
          Options.ocamldep   := ocamlfind & A"ocamldep";
          Options.ocamldoc   := ocamlfind & A"ocamldoc";
          Options.ocamlmktop := ocamlfind & A"ocamlmktop";
          Options.ocamlmklib := ocamlfind & A"ocamlmklib"

      | After_rules ->

          (* When one link an OCaml library/binary/package, one should use
           * -linkpkg *)
          flag ["ocaml"; "link"; "program"] & A"-linkpkg";

          if not (conf.no_automatic_syntax) then begin
            (* For each ocamlfind package one inject the -package option when
             * compiling, computing dependencies, generating documentation and
             * linking. *)
            List.iter
              begin fun pkg ->
                let base_args = [A"-package"; A pkg] in
                (* TODO: consider how to really choose camlp4o or camlp4r. *)
                let syn_args = [A"-syntax"; A "camlp4o"] in
                let (args, pargs) =
                  (* Heuristic to identify syntax extensions: whether they end in
                     ".syntax"; some might not.
                  *)
                  if Filename.check_suffix pkg "syntax" ||
                     List.mem pkg well_known_syntax then
                    (syn_args @ base_args, syn_args)
                  else
                    (base_args, [])
                in
                flag ["ocaml"; "compile";  "pkg_"^pkg] & S args;
                flag ["ocaml"; "ocamldep"; "pkg_"^pkg] & S args;
                flag ["ocaml"; "doc";      "pkg_"^pkg] & S args;
                flag ["ocaml"; "link";     "pkg_"^pkg] & S base_args;
                flag ["ocaml"; "infer_interface"; "pkg_"^pkg] & S args;

                (* TODO: Check if this is allowed for OCaml < 3.12.1 *)
                flag ["ocaml"; "compile";  "package("^pkg^")"] & S pargs;
                flag ["ocaml"; "ocamldep"; "package("^pkg^")"] & S pargs;
                flag ["ocaml"; "doc";      "package("^pkg^")"] & S pargs;
                flag ["ocaml"; "infer_interface"; "package("^pkg^")"] & S pargs;
              end
              (find_packages ());
          end;

          (* Like -package but for extensions syntax. Morover -syntax is useless
           * when linking. *)
          List.iter begin fun syntax ->
          flag ["ocaml"; "compile";  "syntax_"^syntax] & S[A"-syntax"; A syntax];
          flag ["ocaml"; "ocamldep"; "syntax_"^syntax] & S[A"-syntax"; A syntax];
          flag ["ocaml"; "doc";      "syntax_"^syntax] & S[A"-syntax"; A syntax];
          flag ["ocaml"; "infer_interface"; "syntax_"^syntax] &
                S[A"-syntax"; A syntax];
          end (find_syntaxes ());

          (* The default "thread" tag is not compatible with ocamlfind.
           * Indeed, the default rules add the "threads.cma" or "threads.cmxa"
           * options when using this tag. When using the "-linkpkg" option with
           * ocamlfind, this module will then be added twice on the command line.
           *
           * To solve this, one approach is to add the "-thread" option when using
           * the "threads" package using the previous plugin.
           *)
          flag ["ocaml"; "pkg_threads"; "compile"] (S[A "-thread"]);
          flag ["ocaml"; "pkg_threads"; "doc"] (S[A "-I"; A "+threads"]);
          flag ["ocaml"; "pkg_threads"; "link"] (S[A "-thread"]);
          flag ["ocaml"; "pkg_threads"; "infer_interface"] (S[A "-thread"]);
          flag ["ocaml"; "package(threads)"; "compile"] (S[A "-thread"]);
          flag ["ocaml"; "package(threads)"; "doc"] (S[A "-I"; A "+threads"]);
          flag ["ocaml"; "package(threads)"; "link"] (S[A "-thread"]);
          flag ["ocaml"; "package(threads)"; "infer_interface"] (S[A "-thread"]);

      | _ ->
          ()
end

module MyOCamlbuildBase = struct
(* # 22 "src/plugins/ocamlbuild/MyOCamlbuildBase.ml" *)


  (** Base functions for writing myocamlbuild.ml
      @author Sylvain Le Gall
    *)





  open Ocamlbuild_plugin
  module OC = Ocamlbuild_pack.Ocaml_compiler


  type dir = string
  type file = string
  type name = string
  type tag = string


(* # 62 "src/plugins/ocamlbuild/MyOCamlbuildBase.ml" *)


  type t =
      {
        lib_ocaml: (name * dir list * string list) list;
        lib_c:     (name * dir * file list) list;
        flags:     (tag list * (spec OASISExpr.choices)) list;
        (* Replace the 'dir: include' from _tags by a precise interdepends in
         * directory.
         *)
        includes:  (dir * dir list) list;
      }


  let env_filename =
    Pathname.basename
      BaseEnvLight.default_filename


  let dispatch_combine lst =
    fun e ->
      List.iter
        (fun dispatch -> dispatch e)
        lst


  let tag_libstubs nm =
    "use_lib"^nm^"_stubs"


  let nm_libstubs nm =
    nm^"_stubs"


  let dispatch t e =
    let env =
      BaseEnvLight.load
        ~filename:env_filename
        ~allow_empty:true
        ()
    in
      match e with
        | Before_options ->
            let no_trailing_dot s =
              if String.length s >= 1 && s.[0] = '.' then
                String.sub s 1 ((String.length s) - 1)
              else
                s
            in
              List.iter
                (fun (opt, var) ->
                   try
                     opt := no_trailing_dot (BaseEnvLight.var_get var env)
                   with Not_found ->
                     Printf.eprintf "W: Cannot get variable %s\n" var)
                [
                  Options.ext_obj, "ext_obj";
                  Options.ext_lib, "ext_lib";
                  Options.ext_dll, "ext_dll";
                ]

        | After_rules ->
            (* Declare OCaml libraries *)
            List.iter
              (function
                 | nm, [], intf_modules ->
                     ocaml_lib nm;
                     let cmis =
                       List.map (fun m -> (String.uncapitalize m) ^ ".cmi")
                                intf_modules in
                     dep ["ocaml"; "link"; "library"; "file:"^nm^".cma"] cmis
                 | nm, dir :: tl, intf_modules ->
                     ocaml_lib ~dir:dir (dir^"/"^nm);
                     List.iter
                       (fun dir ->
                          List.iter
                            (fun str ->
                               flag ["ocaml"; "use_"^nm; str] (S[A"-I"; P dir]))
                            ["compile"; "infer_interface"; "doc"])
                       tl;
                     let cmis =
                       List.map (fun m -> dir^"/"^(String.uncapitalize m)^".cmi")
                                intf_modules in
                     dep ["ocaml"; "link"; "library"; "file:"^dir^"/"^nm^".cma"]
                         cmis)
              t.lib_ocaml;

            (* Declare directories dependencies, replace "include" in _tags. *)
            List.iter
              (fun (dir, include_dirs) ->
                 Pathname.define_context dir include_dirs)
              t.includes;

            (* Declare C libraries *)
            List.iter
              (fun (lib, dir, headers) ->
                   (* Handle C part of library *)
                   flag ["link"; "library"; "ocaml"; "byte"; tag_libstubs lib]
                     (S[A"-dllib"; A("-l"^(nm_libstubs lib)); A"-cclib";
                        A("-l"^(nm_libstubs lib))]);

                   flag ["link"; "library"; "ocaml"; "native"; tag_libstubs lib]
                     (S[A"-cclib"; A("-l"^(nm_libstubs lib))]);

                   flag ["link"; "program"; "ocaml"; "byte"; tag_libstubs lib]
                     (S[A"-dllib"; A("dll"^(nm_libstubs lib))]);

                   (* When ocaml link something that use the C library, then one
                      need that file to be up to date.
                      This holds both for programs and for libraries.
                    *)
  		 dep ["link"; "ocaml"; tag_libstubs lib]
  		     [dir/"lib"^(nm_libstubs lib)^"."^(!Options.ext_lib)];

  		 dep  ["compile"; "ocaml"; tag_libstubs lib]
  		      [dir/"lib"^(nm_libstubs lib)^"."^(!Options.ext_lib)];

                   (* TODO: be more specific about what depends on headers *)
                   (* Depends on .h files *)
                   dep ["compile"; "c"]
                     headers;

                   (* Setup search path for lib *)
                   flag ["link"; "ocaml"; "use_"^lib]
                     (S[A"-I"; P(dir)]);
              )
              t.lib_c;

              (* Add flags *)
              List.iter
              (fun (tags, cond_specs) ->
                 let spec = BaseEnvLight.var_choose cond_specs env in
                 let rec eval_specs =
                   function
                     | S lst -> S (List.map eval_specs lst)
                     | A str -> A (BaseEnvLight.var_expand str env)
                     | spec -> spec
                 in
                   flag tags & (eval_specs spec))
              t.flags
        | _ ->
            ()


  let dispatch_default conf t =
    dispatch_combine
      [
        dispatch t;
        MyOCamlbuildFindlib.dispatch conf;
      ]


end


# 606 "myocamlbuild.ml"
open Ocamlbuild_plugin;;
let package_default =
  {
     MyOCamlbuildBase.lib_ocaml = [("win-error", ["lib"], [])];
     lib_c = [];
     flags = [];
     includes = [("lib_test", ["lib"])]
  }
  ;;

let conf = {MyOCamlbuildFindlib.no_automatic_syntax = false}

let dispatch_default = MyOCamlbuildBase.dispatch_default conf package_default;;

# 622 "myocamlbuild.ml"
(* OASIS_STOP *)
Ocamlbuild_plugin.dispatch dispatch_default;;
